nvml/datadog_checks/nvml/nvml.py

```diff
@@ -0,0 +1,198 @@
+import pynvml
+
+from datadog_checks.base import AgentCheck
+
+from .api_pb2 import ListPodResourcesRequest
+from .api_pb2_grpc import PodResourcesListerStub
+
+METRIC_PREFIX = "nvml."
+SOCKET_PATH = "/var/lib/kubelet/pod-resources/kubelet.sock"
+"""Assumed to be a UDS accessible from this running code"""
+
+
+class NvmlInit(object):
+    """Wraps an nvmlInit and an nvmlShutdown inside the same context"""
+
+    def __enter__(self):
+        NvmlCheck.N.nvmlInit()
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        NvmlCheck.N.nvmlShutdown()
+
+
+class NvmlCall(object):
+    previously_printed_errors = set()
+    """Wraps a call and checks for an exception (of any type).
+
+       Why this exists: If a graphics card doesn't support a nvml method, we don't want to spam the logs with just
+       that method's errors, but we don't want to never error.  And we don't want to fail all the
+       metrics, just the metrics that aren't supported.  This class supports that use case.
+
+       NvmlCall wraps a call and checks for an exception (of any type).  If an exception is raised
+       then that error is logged, but only logged once for this type of call
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        # Do nothing if the exception is not from pynvml or there is no exception
+        if traceback is None:
+            return False
+        if exception_type is not pynvml.NVMLError:
+            return False
+
+        # Suppress pynvml exceptions so we can continue
+        if self.name in self.previously_printed_errors:
+            return True
+        self.previously_printed_errors.add(self.name)
+        self.log.warning("Unable to execute NVML function: %s: %s", self.name, exception_value)
+        return True
+
+
+class NvmlCheck(AgentCheck):
+    N = pynvml
+    """The pynvml package, explicitly assigned, used for easy test mocking."""
+    known_tags = {}
+    """A map of GPU UUIDs to the k8s tags we should assign that GPU."""
+    lock = threading.Lock()
+    """Lock for the object known_tags."""
+    _thread = None
+    """Daemon thread updating k8s tag information in the background."""
+
+    def add_gauge(self, name, value, tags=None):
+        if tags is None:
+            tags = []
+        self.gauge(METRIC_PREFIX + name, value, tags)
+
+    def add_monotonic_count(self, name, value, tags=None):
+        if tags is None:
+            tags = []
+        self.monotonic_count(METRIC_PREFIX + name, value, tags)
+
+    def check(self, instance):
+        # Start thread once and keep it running in the background
+        if self._thread is None:
+            self._start_discovery()
+        with NvmlInit():
+            self.gather(instance)
+
+    def gather(self, instance):
+        with NvmlCall("device_count"):
+            deviceCount = NvmlCheck.N.nvmlDeviceGetCount()
```

Comment:
Let's instead use `self.` everywhere we can

Reply:
I'm not using self so I can easily stub it out during the test.  For example, object NvmlInit.  I tried a few dozen permutations and cannot find another that works well with mock.